lib/plugins.js

```diff
@@ -70,87 +102,145 @@ Plugins.printPluginResults = function(specResults) {
-            log.puts('Plugin: ' + names[i] + ' (' + funName + ')');
-            Plugins.printPluginResults(pluginResult.specResults);
-          }
+/*
+ * Generates the handler for a plugin function (e.g. the setup() function)
+ *
+ * @param {string} funName The name of the function to make a handler for
+ * @param {boolean=} failReturnVal The value that the function should return if
+ *     the plugin crashes
+ *
+ * @return {Function} The handler
+ */
+function pluginFunFactory(funName, failReturnVal) {
+  return function() {
+    var promises = [];
+    var self = this;
+    var args = arguments;
 
-          // Join objects
-          ret = helper.joinTestLogs(ret, pluginResult);
-        }
+    this.pluginObjs.forEach(function(pluginObj) {
+      if (pluginObj[funName]) {
+        promises.push(safeCallPluginFun(pluginObj, funName, args,
+          self.resultsReported, failReturnVal));
       }
-
-      return ret;
     });
+
+    return q.all(promises);
   };
 }
 
-/**
- * Sets up plugins before tests are run.
+/*
+ * Generates the handler for a plugin function with a return value (e.g. the
+ * rewriteURL() function).
  *
- * @return {q.Promise} A promise which resolves when the plugins have all been
- *     set up.
- */
-Plugins.prototype.setup = pluginFunFactory('setup');
-
-/**
- * Tears down plugins after tests are run.
+ * @param {string} funName The name of the function to make a handler for
  *
- * @return {q.Promise} A promise which resolves when the plugins have all been
- *     torn down.
+ * @return {Function} The handler
  */
-Plugins.prototype.teardown = pluginFunFactory('teardown');
+function pluginMutatorFunFactory(funName) {
+  return function(val) {
+    var result = q(val);
+    var self = this;
+    var args = arguments;
+
+    this.pluginObjs.forEach(function(pluginObj) {
+      if (pluginObj[funName]) {
+        result.then(function(val) {
+          args[0] = val;
+          return safeCallPluginFun(pluginObj, funName, args,
+              self.resultsReported, val);
+        });
+      }
+    });
 
-/**
- * Run after the test results have been processed (any values returned will
- * be ignored), but before the process exits. Final chance for cleanup.
- *
- * @return {q.Promise} A promise which resolves when the plugins have all been
- *     torn down.
- */
-Plugins.prototype.postResults = pluginFunFactory('postResults');
+    return result;
+  };
+}
 
 /**
- * Called after each test block completes.
- *
- * @return {q.Promise} A promise which resolves when the plugins have all been
- *     torn down.
+ * @see docs/plugins.md#writing-plugins for information on these functions
  */
+Plugins.prototype.setup = pluginFunFactory('setup');
+Plugins.prototype.teardown = pluginFunFactory('teardown');
+Plugins.prototype.postResults = pluginFunFactory('postResults');
 Plugins.prototype.postTest = pluginFunFactory('postTest');
+Plugins.prototype.rewriteURL = pluginMutatorFunFactory('rewriteURL');
```

Comment:
Knee-jerk reaction: I dislike the special infrastructure for the mutator, it seems more complicated. Does the URL rewriter belong in plugins? Can that be done in a better way?